subxt/src/events.rs

```diff
@@ -470,152 +798,357 @@ mod tests {
     }
 
     #[test]
-    fn decode_single_event() {
-        #[derive(Clone, Encode, TypeInfo)]
+    fn statically_decode_single_event() {
+        #[derive(Clone, Debug, PartialEq, Decode, Encode, TypeInfo)]
         enum Event {
             A(u8),
         }
 
-        let pallet_index = 0;
-        let pallet = pallet_metadata::<Event>(pallet_index);
-        let decoder = init_decoder(vec![pallet]);
+        // Create fake metadata that knows about our single event, above:
+        let metadata = metadata::<Event>();
 
-        let event = Event::A(1);
-        let encoded_event = event.encode();
-        let event_records = vec![event_record(pallet_index, event)];
+        // Encode our events in the format we expect back from a node, and
+        // construst an Events object to iterate them:
+        let events = events::<Event>(
+            &metadata,
+            vec![event_record(Phase::Finalization, Event::A(1))],
+        );
 
-        let mut input = Vec::new();
-        event_records.encode_to(&mut input);
+        let event_details: Vec<EventDetails<AllEvents<Event>>> =
+            events.iter().collect::<Result<_, _>>().unwrap();
+        assert_eq!(
+            event_details,
+            vec![EventDetails {
+                index: 0,
+                phase: Phase::Finalization,
+                event: AllEvents::E(Event::A(1))
+            }]
+        );
+    }
 
-        let events = decoder.decode_events(&mut &input[..]).unwrap();
+    #[test]
+    fn statically_decode_multiple_events() {
```

Comment:
Would be good to have a test for decoding one or more events successfully, and then encountering one it can't decode.

Reply:
Great idea!